tests/Aql/ExecutionBlockImplTest.cpp

```diff
@@ -1494,6 +1824,477 @@ TEST_P(ExecutionBlockImplExecuteIntegrationTest, test_call_forwarding_implement_
+  AqlCallStack stack{call};
+  if (doesWaiting()) {
+    auto const [state, skipped, block] = testee->execute(stack);
+    EXPECT_EQ(state, ExecutionState::WAITING);
+    EXPECT_EQ(skipped, 0);
+    EXPECT_EQ(block, nullptr);
+  }
+  auto const [state, skipped, block] = testee->execute(stack);
+
+  if (std::holds_alternative<size_t>(call.softLimit) && !call.hasHardLimit()) {
+    EXPECT_EQ(state, ExecutionState::HASMORE);
+    // Do not append shadowRow on softLimit
+    ValidateResult(builder, skipped, block, outReg, 0);
+  } else {
+    EXPECT_EQ(state, ExecutionState::DONE);
+    // Forward to shadowRow on hardLimit
+    ValidateResult(builder, skipped, block, outReg, 1);
+    ASSERT_TRUE(block != nullptr);
+    ValidateShadowRow(block, block->size() - 1, 0);
+  }
+}
+
+// Test a classical input ending in a relevant shadowRow and a non-relevant shadow_row
+TEST_P(ExecutionBlockImplExecuteIntegrationTest, input_and_non_relevant_shadowRow) {
+  std::deque<SharedAqlItemBlockPtr> blockDeque;
+  {
+    SharedAqlItemBlockPtr block = buildBlock<0>(fakedQuery->engine()->itemBlockManager(),
+                                                {{}, {}, {}}, {{1, 0}, {2, 1}});
+    blockDeque.push_back(std::move(block));
+  }
+  auto singleton = std::make_unique<WaitingExecutionBlockMock>(
+      fakedQuery->engine(), generateNodeDummy(), std::move(blockDeque),
+      doesWaiting() ? WaitingExecutionBlockMock::WaitingBehaviour::ALLWAYS
+                    : WaitingExecutionBlockMock::WaitingBehaviour::NEVER);
+
+  auto builder = std::make_shared<VPackBuilder>();
+  {
+    builder->openArray();
+    for (size_t i = 0; i < 998; ++i) {
+      builder->add(VPackValue(i));
+    }
+    builder->close();
+  }
+
+  RegisterId outReg = 0;
+  auto producer = produceBlock(singleton.get(), builder, outReg);
+
+  CallAsserter getAsserter{getCall()};
+  SkipCallAsserter skipAsserter{getCall()};
+  auto testee = forwardBlock(getAsserter, skipAsserter, producer.get(), outReg);
+
+  auto const& call = getCall();
+  AqlCallStack stack{call};
+  if (doesWaiting()) {
+    auto const [state, skipped, block] = testee->execute(stack);
+    EXPECT_EQ(state, ExecutionState::WAITING);
+    EXPECT_EQ(skipped, 0);
+    EXPECT_EQ(block, nullptr);
+  }
+  auto const [state, skipped, block] = testee->execute(stack);
+
+  if (std::holds_alternative<size_t>(call.softLimit) && !call.hasHardLimit()) {
+    EXPECT_EQ(state, ExecutionState::HASMORE);
+    // Do not append shadowRow on softLimit
+    ValidateResult(builder, skipped, block, outReg, 0);
+  } else {
+    EXPECT_EQ(state, ExecutionState::DONE);
+    // Forward to shadowRow on hardLimit
+    ValidateResult(builder, skipped, block, outReg, 2);
+    ASSERT_TRUE(block != nullptr);
+    // Include both shadow rows
+    ValidateShadowRow(block, block->size() - 2, 0);
+    ValidateShadowRow(block, block->size() - 1, 1);
+  }
+}
+
+// Test multiple subqueries
+TEST_P(ExecutionBlockImplExecuteIntegrationTest, multiple_subqueries) {
```

Comment:
After the experience with bugs pertaining to empty inputs and nested subqueries, we should explicitly test empty inputs and nested subqueries, I think.